packages/telemetry/src/flight-recorder.js

```diff
@@ -0,0 +1,249 @@
+  circularBufferSize = DEFAULT_CIRCULAR_BUFFER_SIZE,
+  stateDir = '/tmp',
+  circularBufferFile,
+}) => {
+  const bufferFile =
+    circularBufferFile || `${stateDir}/${DEFAULT_CIRCULAR_BUFFER_FILE}`;
+  // console.log({ circularBufferFile, bufferFile });
+
+  const newArenaSize = await initializeCircularBuffer(
+    bufferFile,
+    circularBufferSize,
+  );
+
+  /** @type {Uint8Array} */
+  const fileBuf = BufferFromFile(bufferFile).Uint8Array();
+  const header = new DataView(fileBuf.buffer, 0, I_ARENA_START);
+
+  // Detect the arena size from the header, if not initialized.
+  const hdrArenaSize = header.getBigUint64(I_ARENA_SIZE);
+  const arenaSize = newArenaSize || hdrArenaSize;
+
+  const hdrMagic = header.getBigUint64(I_MAGIC);
+  assert.equal(
+    SLOG_MAGIC,
+    hdrMagic,
+    X`${bufferFile} is not a slog buffer; wanted magic ${SLOG_MAGIC}, got ${hdrMagic}`,
+  );
+  assert.equal(
+    arenaSize,
+    hdrArenaSize,
+    X`${bufferFile} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`,
+  );
+  const arena = new Uint8Array(
+    fileBuf.buffer,
+    header.byteLength,
+    Number(arenaSize),
+  );
+
+  /**
+   * @param {Uint8Array} data
+   * @param {number} [offset]
+   * @returns {IteratorResult<Uint8Array, void>}
+   */
+  const readCircBuf = (data, offset = 0) => {
+    assert(
+      offset + data.byteLength <= arenaSize,
+      X`Reading past end of circular buffer`,
+    );
+
+    // Read the data to the end of the arena.
+    let firstReadLength = data.byteLength;
+    const circStart = header.getBigUint64(I_CIRC_START);
+    const circEnd = header.getBigUint64(I_CIRC_END);
+    const readStart = (Number(circStart) + offset) % Number(arenaSize);
+    if (circStart > circEnd) {
+      // The data is wrapped around the end of the arena, like BBB---AAA
+      firstReadLength = Math.min(
+        firstReadLength,
+        Number(arenaSize) - readStart,
+      );
+      if (readStart >= circEnd && readStart < circStart) {
+        return { done: true, value: undefined };
+      }
+    } else if (readStart < circStart || readStart >= circEnd) {
+      // The data is contiguous, like ---AAABBB---
+      return { done: true, value: undefined };
+    }
+    data.set(arena.subarray(readStart, readStart + firstReadLength));
+    if (firstReadLength < data.byteLength) {
+      data.set(
+        arena.subarray(0, data.byteLength - firstReadLength),
+        firstReadLength,
+      );
+    }
+    return { done: false, value: data };
+  };
+
+  /** @param {Uint8Array} data */
+  const writeCircBuf = data => {
+    if (BigUint64Array.BYTES_PER_ELEMENT + data.byteLength > arena.byteLength) {
+      // The data is too big to fit in the arena, so skip it.
+      const tooBigRecord = JSON.stringify({
+        type: 'slog-record-too-big',
+        size: data.byteLength,
+      });
+      data = new TextEncoder().encode(tooBigRecord);
+    }
+
+    const record = new Uint8Array(
```

Comment:
Hm, using a separate array to build up the record, then blitting the record into the arena? Not sure that buys much over two writes into the arena, and I wonder about the cost of creating new typed arrays. We don't get any concurrency protection beyond not updating the start pointer until after the record writes have finished (a consumer might sample the file in the middle of a `memcpy`). I'd use a static buffer to do the `u64` conversion, then a pair of writes. (Now I understand what you mentioned about still needing zero- or one-copy optimizations).

Same concern applies to the new buffer we create for each record that needs to be skipped, to make more space.

Reply:
I really want this code to drive from the opposite direction: the caller supplies an iterator of buffers to add to a single record (which we can connect to, say, an 8K buffer textencoder), and this code pulls the data from the iterator into the circular buffer.

The reader would just be an iterator that returns record data buffers.

Reply:
Just a thought (haven't gone through the circular buffer logic). Could we update the begin pointer to remove the chunk about to be overwritten (if applicable), write the chunk, then update the end pointer to let consumers know the chunk has been included? Or did I completely misunderstand the issue?